File: LeetCode/Medium/Daily Temperatures/Go/daily_temperatures.go
package main

import "fmt"

func daily_temperatures(temperatures []int) []int {
	n := len(temperatures)
	res := make([]int, n)
	st := []int{}
	for i := 0; i < n; i++ {
		for len(st) > 0 && temperatures[i] > temperatures[st[len(st)-1]] {
			res[st[len(st)-1]] = i - st[len(st)-1]
			st = st[:len(st)-1]
		}
		st = append(st, i)
	}
	return res
}

func main() {
	fmt.Println(daily_temperatures([]int{73, 74, 75, 71, 69, 72, 76, 73}))
}

File: LeetCode/neetcode 150/stack/daily_temperatures.cc
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        stack<int> st;
        vector<int> res(n);
        st.push(n - 1);
        res[n - 1] = 0;

        for (int i = n - 2; i >= 0; i--) {
            while (!st.empty() && temperatures[i] >= temperatures[st.top()]) {
                st.pop();
            }
            if (st.empty()) {
                res[i] = 0;
            } else {
                res[i] = st.top() - i;
            }
            st.push(i);
        }
        return res;
    }
};
